File: src/main/scala/com/github/vickumar1981/stringdistance/ArrayDistance.scala
package com.github.vickumar1981.stringdistance

import com.github.vickumar1981.stringdistance.impl._

object ArrayDistance {
  sealed trait ScoreMetric {
    def score[T](arr1: Array[T], arr2: Array[T]): Double
  }

  sealed trait DistanceMetric {
    def distance[T](arr1: Array[T], arr2: Array[T]): Int
  }

  sealed trait WeightedScoreMetric[T2] {
    def score[T1](arr1: Array[T1], arr2: Array[T1], weight: T2): Double
    def score[T1](arr1: Array[T1], arr2: Array[T1]): Double
  }

  sealed trait WeightedDistanceMetric[T2] {
    def distance[T1](arr1: Array[T1], arr2: Array[T1], weight: T2): Int
    def distance[T1](arr1: Array[T1], arr2: Array[T1]): Int
  }

  sealed trait ScoreFromDistanceMetric {
    def distance[T](arr1: Array[T], arr2: Array[T]): Int
    private def score[T](s1: Array[T], s2: Array[T]): Double = {
      val maxLen = math.max(s1.length, s2.length)
      val minLen = maxLen - distance(s1, s2)
      (if (minLen < 0 || minLen > maxLen) 0d else minLen * 1d) / maxLen
    }
  }

  object Cosine extends ScoreMetric with CosSimilarityImpl {
    override def score[T](arr1: Array[T], arr2: Array[T]): Double = cosSimilarity(arr1, arr2)
  }

  object Damerau extends ScoreFromDistanceMetric with LevenshteinDistanceImpl {
    override def distance[T](arr1: Array[T], arr2: Array[T]): Int = damerauLevenshtein(arr1, arr2)
  }

  object DiceCoefficient extends ScoreMetric with DiceCoefficientImpl {
    override def score[T](arr1: Array[T], arr2: Array[T]): Double = diceCoefficient(arr1, arr2)
  }

  object Hamming extends ScoreFromDistanceMetric with HammingImpl {
    override def distance[T](arr1: Array[T], arr2: Array[T]): Int = hamming(arr1, arr2)
  }

  object Jaccard extends WeightedScoreMetric[Int] with JaccardImpl {
    override def score[T](arr1: Array[T], arr2: Array[T]): Double = jaccard(arr1, arr2)
    override def score[T](arr1: Array[T], arr2: Array[T], weight: Int): Double = jaccard(arr1, arr2, weight)
  }

  object Jaro extends ScoreMetric with JaroImpl {
    override def score[T](arr1: Array[T], arr2: Array[T]): Double = jaro(arr1, arr2)
  }

  object JaroWinkler extends WeightedScoreMetric[Double] with JaroImpl {
    override def score[T1](arr1: Array[T1], arr2: Array[T1]): Double = jaroWinkler(arr1, arr2)
    override def score[T1](arr1: Array[T1], arr2: Array[T1], weight: Double): Double = jaroWinkler(arr1, arr2, weight)
  }

  object Levenshtein extends ScoreFromDistanceMetric with LevenshteinDistanceImpl {
    override def distance[T](arr1: Array[T], arr2: Array[T]): Int = levenshtein(arr1, arr2)
  }

  object LongestCommonSeq extends DistanceMetric with LongestCommonSeqImpl {
    override def distance[T](arr1: Array[T], arr2: Array[T]): Int = longestCommonSeq(arr1, arr2)
  }

  object NeedlemanWunsch extends WeightedScoreMetric[ConstantGap] with NeedlemanWunschImpl {
    override def score[T1](arr1: Array[T1], arr2: Array[T1]): Double = needleman(arr1, arr2)
    override def score[T1](arr1: Array[T1], arr2: Array[T1], weight: ConstantGap): Double =
      needleman(arr1, arr2, weight)
  }

  object NGram extends WeightedScoreMetric[Int] with WeightedDistanceMetric[Int] with NGramImpl {
    override def distance[T1](arr1: Array[T1], arr2: Array[T1]): Int = nGramDist(arr1, arr2)
    override def distance[T1](arr1: Array[T1], arr2: Array[T1], weight: Int): Int = nGramDist(arr1, arr2, weight)
    override def score[T1](arr1: Array[T1], arr2: Array[T1]): Double = nGram(arr1, arr2)
    override def score[T1](arr1: Array[T1], arr2: Array[T1], weight: Int): Double = nGram(arr1, arr2, weight)

  }

  object Overlap extends WeightedScoreMetric[Int] with OverlapImpl {
    override def score[T1](arr1: Array[T1], arr2: Array[T1]): Double = overlap(arr1, arr2)
    override def score[T1](arr1: Array[T1], arr2: Array[T1], weight: Int): Double =
      overlap(arr1, arr2, weight)
  }

  object SmithWaterman extends WeightedScoreMetric[(Gap, Int)] with SmithWatermanImpl {
    override def score[T1](arr1: Array[T1], arr2: Array[T1]): Double = smithWaterman(arr1, arr2)
    override def score[T1](arr1: Array[T1], arr2: Array[T1], n: (Gap, Int)): Double =
      smithWaterman(arr1, arr2, n._1, n._2)
  }

  object SmithWatermanGotoh extends WeightedScoreMetric[ConstantGap] with SmithWatermanImpl {
    override def score[T1](arr1: Array[T1], arr2: Array[T1]): Double = smithWatermanGotoh(arr1, arr2)
    override def score[T1](arr1: Array[T1], arr2: Array[T1], weight: ConstantGap): Double =
      smithWatermanGotoh(arr1, arr2, weight)
  }

  object Tversky extends WeightedScoreMetric[Double] with JaccardImpl {
    override def score[T1](arr1: Array[T1], arr2: Array[T1]): Double = tversky(arr1, arr2)
    override def score[T1](arr1: Array[T1], arr2: Array[T1], weight: Double): Double =
      tversky(arr1, arr2, weight)
  }

}

File: src/main/java/com/github/vickumar1981/stringdistance/util/ArrayDistance.java
package com.github.vickumar1981.stringdistance.util;

import com.github.vickumar1981.stringdistance.impl.ConstantGap;
import com.github.vickumar1981.stringdistance.impl.Gap;
import com.github.vickumar1981.stringdistance.impl.LinearGap;

import java.util.List;


public class ArrayDistance {
    private final static CosSimilarityImpl cosine = new CosSimilarityImpl();
    private final static JaccardImpl jaccard = new JaccardImpl();
    private final static JaroImpl jaro = new JaroImpl();
    private final static HammingImpl hamming = new HammingImpl();
    private final static DiceCoefficientImpl dice = new DiceCoefficientImpl();
    private final static LevenshteinDistanceImpl levenshtein = new LevenshteinDistanceImpl();
    private final static LongestCommonSeqImpl longestCommonSeq = new LongestCommonSeqImpl();
    private final static NeedlemanWunschImpl needlemanWunsch = new NeedlemanWunschImpl();
    private final static NGramImpl ngram = new NGramImpl();
    private final static OverlapImpl overLap = new OverlapImpl();
    private final static SmithWatermanImpl smithWaterman = new SmithWatermanImpl();

    private static <T> Double scoreFromDistance(T[] s1, T[] s2, Integer distance) {
        int maxLen = Math.max(s1.length, s2.length);
        int minLen = maxLen - distance;
        if (minLen < 0 || minLen > maxLen) {
            minLen = 0;
        }
        return minLen * 1.0 / maxLen;
    }
    
    public static <T> Double cosine(T[] s1, T[] s2) {
        return cosine.cosSimilarity(s1, s2);
    }

    public static <T> Double damerau(T[] s1, T[] s2) {
        return scoreFromDistance(s1, s2, damerauDist(s1, s2));
    }

    public static <T> Integer damerauDist(T[] s1, T[] s2) {
        return levenshtein.damerauLevenshtein(s1, s2);
    }

    public static <T> Double diceCoefficient(T[] s1, T[] s2) {
        return dice.diceCoefficient(s1, s2);
    }

    public static <T> Double hamming(T[] s1, T[] s2) {
        return scoreFromDistance(s1, s2, hammingDist(s1, s2));
    }

    public static <T> Integer hammingDist(T[] s1, T[] s2) {
        return hamming.hamming(s1, s2);
    }

    public static <T> Double jaccard(T[] s1, T[] s2) { return jaccard(s1, s2, 1); }

    public static <T> Double jaccard(T[] s1, T[] s2, Integer n) {
        return jaccard.jaccard(s1, s2, n);
    }

    public static <T> Double jaroWinkler(T[] s1, T[] s2) {
        return jaroWinkler(s1, s2, 0.1);
    }

    public static <T> Double jaroWinkler(T[] s1, T[] s2, Double weight) {
        return jaro.jaroWinkler(s1, s2, weight);
    }

    public static <T> Double jaro(T[] s1, T[] s2) {
        return jaro.jaro(s1, s2);
    }

    public static <T> Double levenshtein(T[] s1, T[] s2) {
        return scoreFromDistance(s1, s2, levenshteinDist(s1, s2));
    }

    public static <T> Integer levenshteinDist(T[] s1, T[] s2) {
        return levenshtein.levenshtein(s1, s2);
    }

    public static <T> Integer longestCommonSeq(T[] s1, T[] s2) {
        return longestCommonSeq.longestCommonSeq(s1, s2);
    }

    public static <T> Double needlemanWunsch(T[] s1, T[] s2) {
        return needlemanWunsch(s1, s2, new ConstantGap(1d, -1d, 0d));
    }

    public static <T> Double needlemanWunsch(T[] s1, T[] s2, ConstantGap gap) {
            return needlemanWunsch.needleman(s1, s2, gap);
    }

    public static <T> Integer nGramDist(T[] s1, T[] s2) { return nGramDist(s1, s2, 1); }

    public static <T> Integer nGramDist(T[] s1, T[] s2, Integer n) {
        return ngram.nGramDist(s1, s2, n);
    }

    public static <T> Double nGram(T[] s1, T[] s2) { return nGram(s1, s2, 1); }

    public static <T> Double nGram(T[] s1, T[] s2, Integer n) {
        return ngram.nGram(s1, s2, n);
    }

    public static <T> List<List<T>> nGramTokens(T[] s1, Integer n) {
        return ngram.tokensAsJava(s1, n);
    }

    public static <T> Double overlap(T[] s1, T[] s2) { return overlap(s1, s2, 1); }

    public static <T> Double overlap(T[] s1, T[] s2, Integer n) {
        return overLap.overlap(s1, s2, n);
    }

    public static <T> Double smithWaterman(T[] s1, T[] s2) {
        return smithWaterman(s1, s2, new LinearGap(1d, -1d, -1d));
    }

    public static <T> Double smithWaterman(T[] s1, T[] s2, Gap gap) {
        return smithWaterman.smithWaterman(s1, s2, gap, Integer.MAX_VALUE);
    }

    public static <T> Double smithWaterman(T[] s1, T[] s2, Gap gap, Integer windowSize) {
        return smithWaterman.smithWaterman(s1, s2, gap, windowSize);
    }

    public static <T> Double smithWatermanGotoh(T[] s1, T[] s2) {
        return smithWatermanGotoh(s1, s2, new ConstantGap(1d, -1d, -1d));
    }

    public static <T> Double smithWatermanGotoh(T[] s1, T[] s2, ConstantGap gap) {
        return smithWaterman.smithWatermanGotoh(s1, s2, gap);
    }

    public static <T> Double tversky(T[] s1, T[] s2) { return tversky(s1, s2, 1d); }

    public static <T> Double tversky(T[] s1, T[] s2, Double n) {
        return jaccard.tversky(s1, s2, n);
    }
}
